pkg/sql/conn_executor.go

```diff
@@ -3113,6 +3128,64 @@ func (ex *connExecutor) handleWaitingForDescriptorIDGeneratorMigration(ctx conte
 	return ex.resetTransactionOnSchemaChangeRetry(ctx)
 }
 
+func (ex *connExecutor) waitOneVersionForNewVersionDescriptorsWithoutJobs(
+	descIDsInJobs catalog.DescriptorIDSet,
+) error {
+	withNewVersion, err := ex.extraTxnState.descCollection.GetOriginalPreviousIDVersionsForUncommitted()
+	if err != nil {
+		return err
+	}
+	for _, idVersion := range withNewVersion {
+		if descIDsInJobs.Contains(idVersion.ID) {
+			continue
+		}
+		if _, err := WaitToUpdateLeases(ex.Ctx(), ex.planner.LeaseMgr(), idVersion.ID); err != nil {
+			// In most cases (normal schema changes), deleted descriptor should have
+			// been handled by jobs. So, normally we won't hit into the situation of
+			// wait for one version of a deleted descriptor. However, we need catch
+			// ErrDescriptorNotFound here because we have a special case of descriptor
+			// repairing where we delete descriptors directly and never record the ids
+			// in jobs payload or details.
+			if errors.Is(err, catalog.ErrDescriptorNotFound) {
+				continue
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func (ex *connExecutor) descIDsInJobs() (catalog.DescriptorIDSet, error) {
```

Comment:
I don't know about this -- this is a very expensive thing to do. 

I think we can do better. Namely, we buffer jobs we create with schema changes for a while. We can extract the desc IDs from that set and we can also extract a set from the declarative schema changer state if we create one, maybe. I'd like it to be the case that we don't go to the kv-store to decide what to wait on.

Reply:
Yeah, I agree. There is sometimes other jobs like query statistics. I think 2 options:
(1) How about we only do this if there is any new version descriptors. This would only affect schema changes with jobs.
(2) Cache job records in the `extraTxnState` as well in addition to the job ids. In legacy schema changer it seems obvious that how job records are passed around. It's not clear to me how we can plumb the job record out of schema changer yet, but I can take a look.

Reply:
For the declarative schema changer, all you need to know is whether a job was created or not.